CONTRACTS: `dfcc_is_cprover_symbol` now uses allow lists to match symbols

remi-delmas-3000 opened the pull request:
Assignments to  special `__CPROVER_*` or `*$object` symbols that are essentially ghost state must not be checked under DFCC.  In the same way built-in `__CPROVER_` or `__VERIFIER` or `nondet_`  functions must not be instrumented by DFCC.

Before, these special symbols were detected by matching a prefix or a suffix, which could allow users to bypass DFCC checks by naming their symbols with these prefixes or suffixes.

Instead we propose to maintain finite sets of function symbols and static symbols which are known to be exempt from DFCC instrumentation, and to test membership in that list. 

We add tests that demonstrate that checks are not bypassed even when symbols have the special prefixes.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

remi-delmas-3000 commented:
If anyone can suggest a better solution (this one is sound but could be hard to maintain as we add new special symbols).

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_is_cprover_symbol.cpp:
```diff
@@ -14,9 +14,142 @@ Date: March 2023
-bool dfcc_is_cprover_symbol(const irep_idt &id)
+#include <unordered_set>
+
+static std::unordered_set<irep_idt> function_symbols;
+static std::unordered_set<irep_idt> static_symbols;
```
Both of these could safely be moved into their respective `dfcc_is_...` functions: the `init` functions would just have to take a reference to the set as a parameter.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_is_cprover_symbol.cpp:
```diff
@@ -14,9 +14,142 @@ Date: March 2023
+{
+  if(static_symbols.empty())
+  {
+    static_symbols.insert(CPROVER_PREFIX "pipe_offset");
+    static_symbols.insert(CPROVER_PREFIX "pipes");
+    static_symbols.insert(CPROVER_PREFIX "rounding_mode");
+    static_symbols.insert(CPROVER_PREFIX "deallocated");
+    static_symbols.insert(CPROVER_PREFIX "dead_object");
+    static_symbols.insert(CPROVER_PREFIX "fpu_control_word");
+    static_symbols.insert(CPROVER_PREFIX "jsa_jump_buffer");
+    static_symbols.insert(CPROVER_PREFIX
+                          "malloc_failure_mode_assert_then_assume");
+    static_symbols.insert(CPROVER_PREFIX "malloc_failure_mode_return_null");
+    static_symbols.insert(CPROVER_PREFIX "malloc_is_new_array");
+    static_symbols.insert(CPROVER_PREFIX "max_malloc_size");
+    static_symbols.insert(CPROVER_PREFIX "memory_leak");
```
Here and above: may I suggest to use lexicographic ordering?

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_is_cprover_symbol.cpp:
```diff
@@ -14,9 +14,142 @@ Date: March 2023
+  init_static_symbols();
   std::string str = id2string(id);
-  return has_prefix(str, CPROVER_PREFIX) || has_prefix(str, "__VERIFIER") ||
-         has_prefix(str, "nondet") || has_suffix(str, "$object");
+  return static_symbols.find(id) != static_symbols.end() ||
+         // auto objects from pointer derefs
+         has_suffix(str, "$object");
```
You will now be better off inlining the `id2string` use so that it will only be done when the lookup in `static_symbols` fails.

tautschnig reviewed (changes requested):
See suggestion below in order to avoid a global-scoped static object.

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_is_cprover_symbol.h:
```diff
@@ -15,8 +15,13 @@ Date: March 2023
+/// `__VERIFIER` or `nondet`.
+bool dfcc_is_cprover_function_symbol(const irep_idt &id);
+
+/// Returns `true` iff the symbol is one of the know CPROVER static
```
```suggestion
/// Returns `true` iff the symbol is one of the known CPROVER static
```

qinheping committed:
CONTRACTS: dfcc_is_cprover_symbol now uses allow lists

The result of `dfcc_is_cprover_symbol` is used to skip frame
conditon checks on symbols with special meaning. Before we
were matching symbols based on their prefix, which would
allow users to skip frame condition checks just by naming their
symbols with these prefixes. We now use a closed allow list,
and distinguish function symbolds form static var symbols.
We add test that show that variables named with __CPROVER,
__VERIFIER or nondet prefixes do not bypass frame condition
checks.
